compiler/vm/vmdef.nim

```diff
@@ -460,77 +470,110 @@ type
 
-  PCtx* = ref TCtx
-  TCtx* = object of TPassContext
-    # XXX: TCtx stores three different things:
-    #  - VM execution state
-    #  - VM environment (code, constants, type data, etc.)
-    #  - vmgen context/state
-    # These should be encapsulated into standalone types
+    currentExceptionA*, currentExceptionB*: HeapSlotHandle
+    exceptionInstr*: int # index of instruction that raised the exception
 
-    code*: seq[TInstr]
-    debug*: seq[TLineInfo]  # line info for every instruction; kept separate
-                            # to not slow down interpretation
-    sframes*: seq[TStackFrame] ## The stack of the currently running code # XXX: rename to `stack`?
-    globals*: seq[HeapSlotHandle] ## Stores each global's corresponding heap slot
-    constants*: seq[VmConstant] ## constant data
+    loopIterations*: int
+
+  ExecStateExtra* = object
+    ## Extra state needed for macros/AST-manipulation that is only
+    ## relevant/available during compile-time execution
+    callsite*: PNode
+    errorFlag*: Report
 
-    typeInfoCache*: TypeInfoCache ## manages the types used by the VM
+    comesFromHeuristic*: TLineInfo # Heuristic for better macro stack traces
 
-    rtti*: seq[VmTypeInfo] ## run-time-type-information as needed by
-                           ## conversion and `repr`
-    functions*: seq[VmFunctionObject] ## all functions known to the VM. Indexed
-                                      ## by `FunctionIndex`
+    module*: PSym
+    cache*: IdentCache
+    graph*: ModuleGraph
+    templInstCounter*: ref int # gives every template instantiation a unique ID, needed here for getAst
+
+  SharedExecState* = object
+    ## State that is shared across VM invocations. Mutated by execution
     memory*: VmMemoryManager
 
+    globals*: seq[HeapSlotHandle] ## Stores each global's corresponding heap slot
+
+    # XXX: the current `const` handling won't work out
     materializedConsts*: BiTable[MaterializedConst] ##
       ## `const` id <-> `const` info
 
-    # XXX: could and should be merged with `procToCodePos` table
-    procToFuncObj*: Table[int, FunctionIndex] ## symbol id -> index into `functions`
+    profiler*: Profiler
 
-    currentExceptionA*, currentExceptionB*: HeapSlotHandle
-    exceptionInstr*: int # index of instruction that raised the exception
-    prc*: PProc
-    module*: PSym
-    callsite*: PNode
+  ExecEnv* = object
+    ## Static state that is shared across VM invocations and is only read
+    ## during execution. It acts as the interface between compile-time and
+    ## run-time
+
+    types*: Types
+    rtti*: seq[VmTypeInfo]
+
+    constants*: seq[VmConstant] ## constant data
+    functions*: seq[VmFunctionObject] ## all functions known to the VM. Indexed
+                                      ## by `FunctionIndex`
+
+    callbacks*: seq[tuple[key: string, value: VmCallback]]
+
+    # XXX: maybe split these off into a separate object (e.g. ExecConfig)?
```

Comment:
I've been reading down the code, so evolving my thinking as I'm going. `Exec` might be good for single execution of the VM?

splitting out ExecConfig makes sense, I think.